File: Two_Pointers_1_Problem_1_sortColors.java
//Time Complexity : O(n)
//Space Complexity : O(1)
//Run successfully on leetcode
//Had to think about which pointers to put condtions on. Understood it perfectly

//Here we will be swaps on the basis of middle pointer. If middle pointer is on its perfect position, will be increasing it
//If middle pointer is behind its position,We will be swapping with low pointer and increasing both the pointers
//If middle pointer is ahead of its position, we will be swapping with high pointer and decreasing the high by one


public class Two_Pointers_1_Problem_1_sortColors {
    public static void sortColors(int[] nums) {
        int length = nums.length;
        int left = 0;
        int mid = 0;
        int right = length - 1;
        while(mid <= right)
        {
            if(nums[mid] == 1)
            {
                mid++;
            }
            else if(nums[mid] == 2)
            {
                swap(nums,mid,right);
                right--;
            }
            else if(nums[mid] == 0)
            {
                swap(nums,mid,left);
                mid++;
                left++;
            }
        }
    }
    public static void swap(int[] nums,int i,int j)
    {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    public static void printArr(int[] nums)
    {
        for(int i =0 ; i < nums.length; i ++)
        {
            System.out.print(nums[i]+ " ");
        }
    }
    public static void main(String[] args) {
        int[] arr = {2,0,2,1,1,0};
        printArr(arr);
        sortColors(arr);
        System.out.println();
        printArr(arr);
    }
}

File: Two-pointers-1 - Problem 1 - sort colors.py
"""
 FAANMG Problem #34 {Medium}

75. Sort Colors

Time Complexity : O(N)

Space Complexity : O(1)


Did this code successfully run on Leetcode : Yes
  

We are using 3 pointers
  
    low=0
    mid =0
    high = last elemnt
    
    loop until mid and high crosses
    
    While iterating mid:
        if mid == 2, then we swap mid and high values and decrementing high pointer
        if mid == 0, then we swap mid and low and then increment both pointer
        
        else if mid == 1 the increment mid pointer

@name: Rahul Govindkumar_RN27JUL2022
"""

class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        
        if(len(nums) ==0):
            return
        
        low = 0
        mid = 0
        high = len(nums) -1
        
        while(mid<= high):
            
            if(nums[mid]==2):
                nums[mid], nums[high] = nums[high], nums[mid]
                high -=1
            
            elif(nums[mid] == 0):
                nums[mid], nums[low] = nums[low], nums[mid]
                low += 1
                mid +=1
            else:
                mid +=1

        